### ISSUE TITLE
a linebreak before an `extends` is a syntax error (but not after)

### ISSUE THREAD

**Issue by dimitropoulos** [microsoft/TypeScript#53589]
# Bug Report

Today, putting a linebreak before the `extends` keyword is a syntax error, but not a line break _after_ the `extends` keyword.

Especially in the context of mapped types which use conditionals (like `Omit`, if you expand it), it can be very surprising if you aren't aware of this.  I can see an argument for leaving it alone in the case of `X5` (below), hopefully the case of `Omit3` demonstrates where this can catch users by surprise.

Perhaps if nothing else, there is some way to improve the error message in this situation (`Omit3` below), which is:

> Mapped object type implicitly has an 'any' template type. (7039)

Which doesn't help as much as the error message in `X5` below:

> Declaration or statement expected. (1128)

### 🔎 Search Terms

line break, linebreak, extends, newline, new line, extends keyword

### 🕗 Version & Regression Information

This is present since at least TypeScript 4.1 when key remapping was introduced.

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about key remapping

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAElByUCEEEGUCiAZAkgOQQLlAOwPYBOAtgIYA2oAzgJ44AuJAHgLABQd1ADgKagAaARgA8AFQB8oALygRobozrccAE0q4ArkQBG3AqAD8oOgXW9sAM3KVuAbjZsOPfgCYp-YQLF3WIUH4B6+vY+YJAA7oQA1mrmAJY4vHR4oAAWJABuvCS43GGgZPFZ5op6AAbyiiqUpQ5cvHwAzKIS0rIVSqpsOJo6eobGpqAWVrbBjvUALG6NQp7evgFBrGy+gDLkoAAKZCQAxvEA5qDZCXkFCaAA+jrmhNwXRim85Qod1aCxatk09ExyBASEWpOPgAVmabhEbHaVQ02l0BiMJjMoEsZGs3iB9QAbNMwXMVmBFsFfJARCkPqAIgRoijCu8cFQ8EReHg6I89Ds8PQKmo2SQ6EcCLwyNxKJ9tLF9uo8OpKJhMaAAPJEWJ0ZyiAA0oAA0lCXjDItxqHhzEccNQWptYjtIprQAhGDsyOplNwhIbjaaRFrtWIvGM6qAAJoCNzK1VCADeR2wAi1WmwrgAvlqAOQkVP+kJ+UCBYlgdZQdQC2Km43qUCy3iq+lHUBC0icTgHLU1yyxNEK8N0JrenVyfWqUAek1mi1uSNsPwAbQ2tZHXqOag2U5z0KH2oRCUyemwGwAutgRLP994kwGnEHXNJu01oyRY-HE6AU6B05n5oTc0t8xBoMX3jLGVK2sd4SwZbIGxIJsWzAlESA7SguxVOgJjtTd1zUBcx0tSdWBnOd4mHI1R1kEhlwHSpOnw-tDG3eE90PGQTzPC9eCDBowxQtD70fUAE1AZM0wzLMFm-NggA)

### 💻 Code

```ts
// ✅ BASELINE: normal syntax
type X1<T> = T extends number ? true : false;

type X2 = X1<1>;
//   ^?

// ✅ works fine to have a new line after `extends`
type X3<T> = T extends
number ? true : false;

type X4 = X3<1>;
//   ^?

// ❌ Placing a new line _before_ the `extends` is a syntax error
type X5<T> = T
extends number ? true : false;

type X6 = X5<1>;
//   ^?

// ✅ This works fine in some other contexts that are less ambiguous:
type Omit2<T, K
extends keyof any> = Pick<T, Exclude<keyof T, K>>;

type Y1 = Omit<{ a: 1, b: 2 }, 'a'>;
//   ^?

// ❌ But if you use it in a remapping, it fails
type Omit3<T, K extends keyof any> = {
  [P in keyof T as P
    extends K ? never : P]: T[P];
}

type Y2 = Omit3<{ a: 1, b: 2 }, 'a'>;
//   ^?


// ✅ But if you use it in a remapping, it fails
type Omit4<T, K extends keyof any> = {
  [P in keyof T as P extends
  K ? never : P]: T[P];
}

type Y3 = Omit4<{ a: 1, b: 2 }, 'a'>;
//   ^?
```

### 🙁 Actual behavior

- `X4` is not a syntax error: you _can_ linebreak _after_ an `extends` keyword in simple conditional types
- `X5` is a syntax error: you cannot linebreak before an `extends` keyword in simple conditional types
- `Omit3` is a syntax error: you cannot linebreak before an `extends` keyword in key remapping
- `Omit4` is not a syntax error: you _can_ linebreak _after_ an `extends` keyword in key remapping

### 🙂 Expected behavior

`Omit3` is the one that I care about the most because it's the one that comes up the most when writing complex types.

here are some more real-world examples where this kind of thing is likely to come up:

<details>

<summary>here are some more real-world examples where this kind of thing is likely to come up</summary>

in these situations, it's pretty reasonable to want to put the `extends` on the next line.
```ts
type GreaterThanDigit<
  First extends Numeric,
  Second extends Numeric
> =
  `0123456789` extends // here
    `${string}${Second}${string}${First}${string}`
  ? true
  : false
```
  
```ts
type UnionToFnInsertion<T> =
  (T extends T ? (arg: () => T) => unknown : never) extends // here
    (arg: infer R) => unknown
  ? R
  : never
```
  
```ts
type Equal<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends // here
  (<T>() => T extends Y ? 1 : 2)
  ? true
  : false
```

</details>

**Comment by dimitropoulos** [comment]
@fatcerberus thanks!  could you please help me understand what's different about

```ts
type Omit2<T, K
extends keyof any> = Pick<T, Exclude<keyof T, K>>;
```

such that it works?

**Comment by dimitropoulos** [comment]
right, so wouldn't that same thing be true for the square bracket starting the map key here:?

```ts
type Omit3<T, K extends keyof any> = {
//                                   ^ Error: Mapped object type implicitly has an 'any' template type. (7039)
  [P in keyof T as P
  ```

**Comment by RyanCavanaugh** [comment]
Some of this is working as intended (X5 is a working-as-intended ASI for example) but this one is a bug
```ts
type M<T> = {
    [K in keyof T as K
    extends "" ? K : K]: any;
}
```

**Comment by Josh-Cena** [comment]
I assume these should be valid too? (A few random places where `parseType` got called)

```ts
type X1<A> = `${A
extends A ? A : A}`;

type X2<A> = X1<A
extends A ? A : A>;

type X3<A, K extends A
extends A ? A : A> = any;

type X4<A, K = A
extends A ? A : A> = any;

type X5<A> = {
  foo: A
}
```

What about these?

```ts
type X5<A> = {
  foo: A
  extends A ? A : A;
};

declare function foo<A, K>(a: K
extends A ? A : A): any;
```
